### PR TITLE
Support NHWC for LRN and batch normalization, and optimize batch normalization speed

### PR DESCRIPTION
**Pull request by yiheng**
## What changes were proposed in this pull request?
Support NHWC for LRN and batch normalization, and optimize batch normalization speed

## How was this patch tested?
unit test. new and existed

### CODE REVIEW

**Comment by yiheng** [comment]
PR valdiation pass

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNorm.scala [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+import com.intel.analytics.bigdl.nn.SpatialBatchNormalization
+import com.intel.analytics.bigdl.nn.abstractnn.DataFormat
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Table
+
+import scala.reflect.ClassTag
+
+class FusedBatchNorm[T: ClassTag](
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNorm.scala]: scala doc

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNorm.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  override def updateOutput(input: Table): Table = {
+    val x = input[Tensor[Float]](1)
+    val scale = input[Tensor[Float]](2)
+    val offset = input[Tensor[Float]](3)
+    val mean = input[Tensor[Float]](4)
+    val variance = input[Tensor[Float]](5)
+
+    if (output.length() == 0) {
+      output(1) = Tensor[Float]().resizeAs(x) // y
+      output(2) = Tensor[Float](x.size(4)) // batch mean
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNorm.scala]: dataformat NHWC and NCHW?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNormGrad.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+import com.intel.analytics.bigdl.nn.{BatchNormalization, SpatialBatchNormalization}
+import com.intel.analytics.bigdl.nn.abstractnn.DataFormat
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Table
+
+import scala.reflect.ClassTag
+
+class FusedBatchNormGrad[T: ClassTag](
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNormGrad.scala]: scala doc?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNormGrad.scala [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  override def updateOutput(input: Table): Table = {
+    val gradOutput = input[Tensor[Float]](1)
+    val x = input[Tensor[Float]](2)
+    val scale = input[Tensor[Float]](3)
+    val saveMean = input[Tensor[Float]](4)
+    val saveVar = input[Tensor[Float]](5)
+
+    if (output.length() == 0) {
+      output(1) = Tensor[Float]().resizeAs(x) // gradInput
+      output(2) = Tensor[Float](x.size(4)) // weight gradient
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNormGrad.scala]: dataformat?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/LRNGrad.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+package com.intel.analytics.bigdl.nn.ops
+
+import com.intel.analytics.bigdl.nn.SpatialCrossMapLRN
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Table
+
+import scala.reflect.ClassTag
+
+class LRNGrad[T: ClassTag](
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/LRNGrad.scala]: scala doc

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/loaders/LRN.scala [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    val size = getInt(nodeDef.getAttrMap, "depth_radius")
+    val k = getFloat(nodeDef.getAttrMap, "bias")
+    val alpha = getFloat(nodeDef.getAttrMap, "alpha")
+    val beta = getFloat(nodeDef.getAttrMap, "beta")
+    SpatialCrossMapLRN[T](
+      size = size * 2 + 1,
+      k = k,
+      alpha = alpha * (size * 2 + 1),
+      beta = beta,
+      format = DataFormat.NHWC
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/loaders/LRN.scala]: only NHWC in LRN?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala [hunk 7]
```diff
@@ -127,545 +124,15 @@ class BatchNormalization[T: ClassTag](
...
-
-        if (null != gradBias && scaleB != 0) {
-          gradBias(_f - 1 + gradBiasOffset) += scaleB * sum
-        }
-      })
-      f += 1
-    }
-    Engine.model.sync(results)
-  }
+  protected val gMean = Tensor[T]()
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: override clearState() ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNorm.scala (hunk 2 above)
Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNorm.scala]: It should be aligned with the dataFormat parameter in the constructor. The default value is NHWC.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNormGrad.scala (hunk 4 above)
Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/FusedBatchNormGrad.scala]: see the above reply

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/loaders/LRN.scala (hunk 6 above)
Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/loaders/LRN.scala]: TF only support NHWC in its definiation

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala (hunk 7 above)
Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: I can see in the original impl there's no clearState(). Not sure if we should clear the runningMean and runningVar as it is useful in the prediction

Comment (yiheng) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BatchNormalization.scala]: clean gMean and gxMean buffer

**Comment by yiheng** [comment]
jenkins pass
